mesop/bin/bin.py

```diff
@@ -84,37 +85,113 @@ def main(argv: Sequence[str]):
   app.run()
 
 
-def fs_watcher(absolute_path: str):
-  """
-  Naive filesystem watcher for the main module of the Mesop application.
+app_modules: set[str] = set()
 
-  In the future, we can watch for the entire directory, but this captures the main 80% case.
-  """
-  last_modified = os.path.getmtime(absolute_path)
-  while True:
-    # Get the current modification time
-    current_modified = os.path.getmtime(absolute_path)
-
-    # Compare the current modification time with the last modification time
-    if current_modified != last_modified and not is_processing_request():
-      # Update the last modification time
-      last_modified = current_modified
+
+def clear_app_modules() -> None:
+  for module in app_modules:
+    # Not every module has been loaded into sys.modules (e.g. the main module)
+    if module in sys.modules:
+      del sys.modules[module]
+
+
+def add_app_module(workspace_dir_path: str, app_module_path: str) -> None:
+  module_name = get_app_module_name(
+    workspace_dir_path=workspace_dir_path, app_module_path=app_module_path
+  )
+  app_modules.add(module_name)
+
+
+def remove_app_module(workspace_dir_path: str, app_module_path: str) -> None:
+  module_name = get_app_module_name(
+    workspace_dir_path=workspace_dir_path, app_module_path=app_module_path
+  )
+  app_modules.remove(module_name)
+
+
+def get_app_module_name(workspace_dir_path: str, app_module_path: str) -> str:
+  relative_path = os.path.relpath(app_module_path, workspace_dir_path)
+  return relative_path.replace(os.sep, ".").rstrip(".py")
+
+
+class ReloadEventHandler(FileSystemEventHandler):
+  def __init__(self, absolute_path: str, workspace_dir_path: str):
+    self.absolute_path = absolute_path
+    self.workspace_dir_path = workspace_dir_path
+
+  def on_any_event(self, event: FileSystemEvent):
+    if not event.is_directory:
       try:
         print("Hot reload: starting...")
         reset_runtime()
-        execute_main_module(absolute_path=absolute_path)
+        execute_main_module(absolute_path=self.absolute_path)
         hot_reload_finished()
         print("Hot reload: finished!")
       except Exception as e:
         logging.log(
           logging.ERROR, "Could not hot reload due to error:", exc_info=e
         )
 
-    time.sleep(0.1)
+  def on_created(self, event: FileSystemEvent):
+    src_path = cast(str, event.src_path)
+    if src_path.endswith(".py"):
+      print(f"Watching new Python module: {event.src_path}")
+      add_app_module(
+        workspace_dir_path=self.workspace_dir_path,
+        app_module_path=event.src_path,
+      )
+
+  def on_deleted(self, event: FileSystemEvent):
+    src_path = cast(str, event.src_path)
+    if src_path.endswith(".py"):
+      print(f"Stopped watching deleted Python module: {event.src_path}")
+      remove_app_module(
+        workspace_dir_path=self.workspace_dir_path,
+        app_module_path=event.src_path,
+      )
+
+
+def fs_watcher(absolute_path: str):
```

Comment:
[question] So absolute path is the full path to the "main.py" file?

Reply:
yup